pkg/resource/deploy/source_eval.go

```diff
@@ -2306,27 +2306,46 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 		}
 		options.DeleteBeforeReplace = opts.DeleteBeforeReplace
 
-		constructResult, err := provider.Construct(ctx, plugin.ConstructRequest{
-			Info:    rm.constructInfo,
-			Type:    t,
-			Name:    name,
-			Parent:  parent,
-			Inputs:  props,
-			Options: options,
-		})
-		if err != nil {
-			var rpcError error
-			rpcError, ok := rpcerror.FromError(err)
-			if !ok {
-				rpcError = err
+		// Run construct in a go routine so we can react to a cancellation on rm.cancel.
+		var constructResult plugin.ConstructResult
+		constructDone := make(chan error)
```

Comment:
I feel we should be able to write a lifecycle test for this, a purposefully bad provider that just hangs in construct after a sub-resource fails and check the deployment still exits ok.

Reply:
Good idea, done in https://github.com/pulumi/pulumi/pull/19976/commits/909b7e911b24d0f982e7e35f52ce1492affc9311